scripts: Fix twisterlib for ruff

LukaszMrugala opened the pull request:
Twisterlib files are often changed and having them under ruff's linting would be useful.
This brings them up to ruff's standards and updates relevant tests.

pdgendt commented:
I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

LukaszMrugala commented:
> I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

Whoops. Didn't really check the diff size while in-progress. When doing the first few files, it seemed unnecessary, as in practice most commits would be for a single file. But now, with the diff size ballooning to c. 3k LOC, I think separating them would be prudent, even if just to make the review easier.

I mostly pushed it here to verify blackboxes in CI and will start the separation after fixing them, if necessary.

golowanow commented:
I'd suggest to split test changes from code changes to reduce risk of regressions when both are changes simultaneously,
IMO better to ruff Twister tests in a separate PR first.

nashif commented:
> I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

@LukaszMrugala  i think @pdgendt  meant a commit per "violation", not a commit per file :-)

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - normalisation

ruff check still detects a few linting errors
inside twisterlib.
We add them to the exclusion list to fix them
issue-by-issue in further commits.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - E402

This fixes errors E402 - module import
not at top of file.
One error in testplan.py left in,
as it is a part of a obscure 2020 fix
that would need to be re-tested.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP035

This fixes errors of type UP035 - deprecated import.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - I001

This fixes the error I001 - unsorted imports.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM118

This fixes the ruff linting error SIM118 - unnecessary
use of keys().

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM401

This fixes the ruff linting error SIM401,
where if-else construction is used instead
of dict.get().

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM300

This fixes ruff linting error SIM300,
where constant values are on the left
side of an expression.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM202, SIM201

This fixes ruff linting error SIM201,
where not is used instead of a more
appropriate comparison operator.
SIM202 exclusion was removed, as it
didn't apply to anything.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM114

This fixes ruff linting error SIM114,
where there are multiple arms in an
if-elif-else construct with the same result.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM110

This fixes ruff linting error SIM1101,
where a builtin, like any() or all()
has been reimplemented.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - E741

This fixes ruff linting error E741,
where a variable name is ambiguous
or hard to read (e.g. l).

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - E713

This fixes ruff linting error E713,
where `not x in y` is used instead
of `x not in y`.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP006

this fixes ruff linting error UP006,
where old typing types are still used
instead of newer, PEP585-compliant ones.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP004

This fixes ruff linting error UP004,
where a redundant `object` inheritance
is present.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP021, UP022

This fixes ruff linting errors UP021 and UP022,
where obsolete subprocess parameters were
still used.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP008

This fixes ruff linting error UP008,
where super() call has unnecessary
parameters.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM102

This fixes ruff linting error SIM102,
where multiple ifs were used instead
of combining them with and or similar
operators.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B009

This fixes ruff linting error B009,
where getattr() was used with a
constant field name.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B905

This fixes ruff linter error B905,
where zip lacks explicit use of the
strict parameter.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B904

This fixes ruff linting error B904,
where an error raised inside an except
clause does not specify from where it
should start.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B023

This fixes ruff linting error B023,
where function definitions use
loop variables incorrectly.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B006

This fixes ruff linting error B006,
where mutables were used as
default parameters.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM105

This fixes ruff linting error SIM105,
where try-except-pass construct was used
instead of contextlib.suppress().

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - B028

This fixes ruff linting error B028,
where warnings.warn() lacked explicit
stacklevel parameter.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP015

This fixes ruff linting error UP015,
where open() has specified unnecessary modes.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - SIM115

This fixes ruff linting error SIM115,
where files were opened without the use of
a context handler.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - F811

This fixes ruff linting error F811,
where a module is reimported or otherwise
redefined without being used.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - F541

This fixes ruff linting error F541,
where an fstring is used when a normal
string would suffice.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP030, UP031, UP032

Fixes ruff linting errors UP030, UP031, UP032,
which make you use format specifiers (fstrings)
instead of printf formatting or str.format().

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - UP007

This fixes ruff linting error UP007,
where one would use Optional or Union
instead of X | Y notation.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Fix twisterlib for ruff - E501

This fixes ruff linting error E501,
indicating overlong lines.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala committed:
scripts: Remove useless PyLint suppression

Consider using with suppression became
obsolte and should be removed.

Signed-off-by: Lukasz Mrugala <lukaszx.mrugala@intel.com>

LukaszMrugala commented:
> I'd suggest to split test changes from code changes to reduce risk of regressions when both are changed simultaneously, IMO better to ruff Twister tests in a separate PR first.

I do not agree with doing the tests first. Currently, before we change tests, they serve as an indicator when changing the main Twister code. If we change tests first, we introduce uncertainty into our twisterlib changes for no reason.

Test changes should be coupled with the twisterlib changes that necessitate them, so no commit introduces broken code (code that fails tests).

golowanow commented:
> I do not agree with doing the tests first. Currently, before we change tests, they serve as an indicator when changing the main Twister code. If we change tests first, we introduce uncertainty into our twisterlib changes for no reason.

This ruff change in Twister is a sort of refactoring without functionality changes, right ?
When you change both Twister tests and Twister code in the same (this) PR, there is risk of false positive (a bug in both simultaneously), so it is less secure.

pdgendt reviewed (commented):
Nice work!

LukaszMrugala commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -123,8 +123,10 @@ def _verify_ztest_suite_name(self, harness_status, detected_suite_names, handler
+            set(_d_suite) != set(expected_suite_names)
+            and not set(_d_suite).issubset(set(expected_suite_names))
+        ):
                 self._missing_suite_name(expected_suite_names, handler_time)
```
Good catch, not sure why linter didn't get it.
It also seemed to miss some %-based formatting, but I've noticed that in time.

LukaszMrugala commented:
> This ruff change in Twister is a sort of refactoring without functionality changes, right ? When you change both Twister tests and Twister code in the same (this) PR, there is risk of false positive (a bug in both simultaneously), so it is less secure.

It should be just refactoring, but some changes are not NOPs. Take for example B006 - mutable argument default. To fix it, you need to change the code. Not to mention the biggest culprit - using context managers for file handling, which completely dissolved two whole functions.

Additionally, unit tests often directly mock functions used by the function under test. If I change that function, how would you have it so both the test passes and is not changed?

pdgendt commented:
Needs a rebase to resolve conflict

glenn-andrews commented:
This breaks testing on my on my Windows installation:

```
west twister --device-testing --device-serial COM3 --device-serial-baud 115200 -p disco_l475_iot1  -T .
Traceback (most recent call last):
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 546, in __call__
    mod = _commands_module_from_file(self.py_file)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 701, in _commands_module_from_file
    spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\west_commands\twister_cmd.py", line 21, in <module>
    from twisterlib.twister_main import main
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\pylib\twister\twisterlib\twister_main.py", line 20, in <module>
    from twisterlib.runner import TwisterRunner
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\pylib\twister\twisterlib\runner.py", line 26, in <module>
    from domains import Domains
ModuleNotFoundError: No module named 'domains'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 505, in run_command
    self.run_extension(args.command, argv)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 623, in run_extension
    self.cmd = self.extensions[name].factory()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 548, in __call__
    raise ExtensionCommandError(
west.commands.ExtensionCommandError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\glenn\zephyrproject\.venv\Scripts\west.exe\__main__.py", line 7, in <module>
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 1085, in main
    app.run(argv or sys.argv[1:])
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 244, in run
    self.run_command(argv, early_args)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 549, in run_command
    self.handle_extension_command_error(ece)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 412, in handle_extension_command_error
    self.cmd.err(msg, fatal=True)
    ^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'err'
```
